keps/sig-instrumentation/1668-trace-context-propagation/README.md

```diff
@@ -0,0 +1,568 @@
+In OpenTelemetry's Go implementation,  span context is passed down through Go context. This will necessitate the threading of context across more of the Kubernetes codebase, which is a [desired outcome regardless](https://github.com/kubernetes/kubernetes/issues/815). In alpha stage,  we need to change some APIs by adding `ctx context.Context` to parameter list whose parameters hasn't contained context.Context yet. Below APIs will be impacted so far.
+
+| APIs                          | file name                                                    |
+| ----------------------------- | ------------------------------------------------------------ |
+| createPods()                  | pkg/controller/controller_utils.go                           |
+| CreatePodsWithControllerRef() | pkg/controller/controller_utils.go<br />pkg/controller/replication/conversion.go<br />pkg/controller/daemon/daemon_controller.go<br />pkg/controller/replication/conversion.go |
+
+
+### Out-of-tree changes
+
+#### Mutating webhook
+
+This webhook will be configured to allow the requests writing the object resource only. For example, the end user create a deployment including 2 replisects, some requests reaching APIServer in order may include:
+Deployment example:
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: sleep
+  namespace: sleeping
+spec:
+  replicas: 2
+  selector:
+    matchLabels:
+      app: sleep
+  template:
+    metadata:
+      labels:
+        app: sleep
+    spec:
+      containers:
+      - name: sleep
+        image: busybox
+        command: ["/bin/sleep","infinity"]
+        imagePullPolicy: IfNotPresent
+```
+Webhook configuration example(read [webhook configuration](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-configuration) for more details):
+```yaml
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  name: trace-context-injector-webhook-cfg
+webhooks:
+  - name: trace-context-injector.k8s.io
+    clientConfig:
+      service:
+        name: trace-context-injector-webhook-svc
+        namespace: default
+        path: "/mutate"
+      caBundle: ${CA_BUNDLE}
+    rules:
+      - operations: ["CREATE","UPDATE"]
+        apiGroups: ["*"]
+        apiVersions: ["*"]
+        resources: ["deployments","deamonsets","replicasets","statefulsets","pods"]
+```
+Requests reaching APIServer:
+
+1. Create deployment: sleep
+2. Create replicaset: sleep-replica
+3. Create 1st pod: sleep-pod-1st
+4. Update 1st pod's subresource: sleep-pod-1st/status
+5. Create 2nd pod: sleep-pod-2nd
+6. Update 2nd pod's subresource: sleep-pod-2nd/status
+7. Update replicaset's subresource: sleep-replica/status
+8. Update deployment's subresource: sleep/status
+
+With above webhook configuration, requests 1, 2, 3, 5 above will reach the webhook and annotations of these object will be updated accordingly. The rest requests will be filtered out in APIServer.
```

Comment:
I read through the `with Mutating webhook` and `without Mutating webhook` cases below and still not sure if the individual controllers can do this task and it will reduce a huge dependency. 
As webhook will be yet another component that will need to be traced to check if adding latency.

Reply:
Sorry for that these sections are out of date. We are trying to use managedFields instead of annotation, which is not using webhook. About why we have to use managedFields, you can find the details begin with [this comment](https://github.com/kubernetes/enhancements/pull/2312#issuecomment-784285072).